libsolidity/analysis/DeclarationContainer.cpp

```diff
@@ -164,3 +184,19 @@ vector<ASTString> DeclarationContainer::similarNames(ASTString const& _name) con
 
 	return similar;
 }
+
+void DeclarationContainer::populateHomonyms(back_insert_iterator<Homonyms> _it) const
```

Comment:
This could also just be
```suggestion
Homonyms DeclarationContainer::homonyms() const
```

Reply:
In an intermediate implementation, I had something like public `Homonyms homonyms()` wrapping private `void populateHomonyms(....)`. That makes the interface cleaner, but requires an extra wrapper, _which is used just once_. So after some thoughts I decided against it. (No strong opinion here though, can go back to it.)

The root cause is that I do not see an easy way to avoid recursive populating of `Homonyms` list.

Reply:
I don't understand why you would need a wrapper, but I also don't understand why recursion would be an issue here. Why doesn't the following work:
```suggestion
Homonyms DeclarationContainer::homonyms() const
{
	Homonyms ret;
	for (DeclarationContainer const* innerContainer: m_innerContainers)
		ret += innerContainer->homonyms();

	for (auto [name, location]: m_homonymCandidates)
		for (auto const* declaration: m_enclosingContainer->resolveName(name, true, true))
			ret.emplace_back(location, declaration);
	return ret;
}
```

Reply:
> I also don't understand why recursion would be an issue here.

Sorry, I did not mention complexity. Let me rephrase:

> The root cause is that I do not see an easy way to avoid recursive populating of Homonyms list
> **which would give `O(N)` complexity**.

Even if quadratic time is not a practical issue here, it does not feel quite right :).

Reply:
I would say that readability always trumps efficiency unless it is proven that the runtime is actually an issue, but we can also keep it like that.

Reply:
Sure, we should be aware of preliminary optimization. But in this concrete case, the code with quadratic complexity might disturb me more, making to think if it was really intended.